lib/common/common/src/tar_ext.rs

```diff
@@ -325,38 +177,97 @@ impl<W: Send + Write + Seek + 'static> BuilderExt<W> {
     }
 
     /// Finish writing the tar archive.
-    pub async fn finish(self) -> std::io::Result<()> {
+    pub async fn finish(self) -> io::Result<()> {
         tokio::task::spawn_blocking(move || self.blocking_finish()).await?
     }
 
     async fn run_async<T, E>(
         &self,
-        f: impl FnOnce(&mut tar::Builder<FusedWriteSeek<W>>) -> Result<T, E> + Send + 'static,
+        f: impl FnOnce(&mut tar::Builder<W>) -> Result<T, E> + Send + 'static,
     ) -> Result<T, E>
     where
         T: Send + 'static,
-        E: Send + 'static + From<JoinError>,
+        E: From<JoinError> + Send + 'static,
     {
-        let tar = Arc::clone(&self.tar);
-        tokio::task::spawn_blocking(move || f(tar.blocking_lock().tar())).await?
+        let tar = self.tar.clone();
+        tokio::task::spawn_blocking(move || f(&mut tar.blocking_lock())).await?
     }
 }
 
-fn join_relative(base: &Path, rel_path: &Path) -> std::io::Result<PathBuf> {
+fn join_relative(base: &Path, rel_path: &Path) -> io::Result<PathBuf> {
     if rel_path.is_absolute() {
-        return Err(std::io::Error::new(
-            std::io::ErrorKind::InvalidInput,
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
             format!("path must be relative, but got {rel_path:?}"),
         ));
     }
 
     Ok(base.join(rel_path))
 }
 
+/// A wrapper that drops inner type `T` inside [`tokio::task::block_in_place`].
+///
+/// This is a is a workaround for the inconvenient [`tar::Builder`] behavior: dropping a [`tar::Builder`]
+/// might cause a final write of archive footer. This behavior is problematic for [`Write`]
+/// implementations that could panic when used in an async context, such as [`SyncIoBridge`].
+///
+/// Note, that [`tokio::task::block_in_place`] **panics** when called under "current-thread" Tokio runtime,
+/// and so this wrapper only works under "multi-threaded" runtime.
+struct BlockInPlaceOnDrop<T>(Option<T>);
```

Comment:
Would it be worth to put this in a separate module, because it's a generic type that may be used elsewhere?

Reply:
Decided to remove this change from the PR.